### PR TITLE
Cleanup of the `state-machine` crate

### PR DESCRIPTION
**Pull request by bkchr**
- Moves `Externalities` and `CodeExecutor` trait so `substrate-primitives`
- Creates a separate file for the state-machine error and implements the trait automatically for all matching types
- Move crate function `new` to the `StateMachine` type itself. 
- `ChildStorageKey` and `NeverOffchainExt` is moved to `substrate-primitives`

I started with this for another pr, but then realized that I probably not need it directly and it is easier to review in a separate pr.

In a future pr, the externalities implementations could probably be moved to its own crate as well.

### CODE REVIEW

File: core/state-machine/src/lib.rs [hunk 1]
```diff
@@ -639,49 +330,57 @@ impl<'a, H, N, B, T, O, Exec> StateMachine<'a, H, N, B, T, O, Exec> where
...
-					let (result, _, storage_delta, changes_delta) = self.execute_aux(compute_tx, false, native_call);
-					(result, storage_delta, changes_delta)
+					let res = self.execute_aux(compute_tx, false, native_call);
+					(res.0, res.2, res.3)
 				},
 				ExecutionManager::NativeWhenPossible => {
-					let (result, _was_native, storage_delta, changes_delta) = self.execute_aux(compute_tx, true, native_call);
-					(result, storage_delta, changes_delta)
+					let res = self.execute_aux(compute_tx, true, native_call);
+					(res.0, res.2, res.3)
```
Comment (pepyakin) [inline core/state-machine/src/lib.rs]: Doesn't look like improvement to me. Shouldn't we just introduce a nominal type for this?

**Comment by pepyakin** [comment]
Can you give a bit more background why, in your opinion, `core/primitives` is a better place for `Externalities`?

**Comment by bkchr** [comment]
1. I'm not quite happy with it in `primitives` as well, but I think it is way better than having it in `state-machine`.
2. `Externalities` is a trait that is used in a lot of places in Substrate and that is the reason why I "promoted" it into `primitives`.
3. As already written in the pr description, I would like to move the implementations we provide for `Externalities` into its own crate. This crate could probably also provide a home for the trait itself :)

File: core/state-machine/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+
+/// State Machine Error bound.
+///
+/// This should reflect WASM error type bound for future compatibility.
```
[the remark this answers is not in the document]
Comment (bkchr) [reply core/state-machine/src/error.rs]: IDK :D I just copied it ^^

File: core/state-machine/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (gnunicorn) [inline core/state-machine/src/error.rs]: Missing file wide docs.
```suggestion

/// State Machine Errors

```

File: core/state-machine/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+pub trait Error: 'static + fmt::Debug + fmt::Display + Send {}
+
+impl<T: 'static + fmt::Debug + fmt::Display + Send> Error for T {}
+
+/// Externalities Error.
+///
+/// Externalities are not really allowed to have errors, since it's assumed that dependent code
+/// would not be executed unless externalities were available. This is included for completeness,
+/// and as a transition away from the pre-existing framework.
+#[derive(Debug, Eq, PartialEq)]
```
Comment (gnunicorn) [inline core/state-machine/src/error.rs]: From the comment I feel this should be marked as soon-to-be-deprecated?

File: core/state-machine/src/lib.rs (hunk 1 above)
Comment (gnunicorn) [reply core/state-machine/src/lib.rs]: Not sure this needs to be done in this PR. We can also put this into a nice starter-ticket.

File: core/state-machine/src/error.rs (hunk 4 above)
Comment (bkchr) [reply core/state-machine/src/error.rs]: I did not add this comment, I just copied it^^ I'm not sure anyone is working on this.

File: core/state-machine/src/lib.rs (hunk 1 above)
Comment (pepyakin) [reply core/state-machine/src/lib.rs]: Good idea1

**Pull request merged**